core/node/storage/pg_stream_trimmer.go

```diff
@@ -0,0 +1,237 @@
+
+	return st, nil
+}
+
+// monitorWorkerPool monitors the worker pool and handles shutdown
+func (t *streamTrimmer) monitorWorkerPool(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			t.log.Info("Worker pool stopped due to context cancellation")
+			t.workerPool.StopWait()
+			return
+		case <-t.stop:
+			t.log.Info("Worker pool stopped due to stop signal")
+			t.workerPool.StopWait()
+			return
+		}
+	}
+}
+
+// processTrimTask processes a single trim task
+func (t *streamTrimmer) processTrimTask(task trimTask) error {
+	ctx, cancel := context.WithTimeout(t.ctx, time.Minute)
+	defer cancel()
+
+	return t.store.txRunner(
+		ctx,
+		"streamTrimmer.processTrimTask",
+		pgx.ReadWrite,
+		func(ctx context.Context, tx pgx.Tx) error {
+			return t.processTrimTaskTx(ctx, tx, task)
+		},
+		nil,
+		"streamId", task.streamId,
+	)
+}
+
+// processTrimTaskTx processes a single trim task within a transaction
+func (t *streamTrimmer) processTrimTaskTx(
+	ctx context.Context,
+	tx pgx.Tx,
+	task trimTask,
+) error {
+	// Get the last snapshot miniblock number
+	lastSnapshotMiniblock, err := t.store.lockStream(ctx, tx, task.streamId, true)
+	if err != nil {
+		return err
+	}
+
+	// Calculate the miniblock number to keep
+	miniblockToKeep := lastSnapshotMiniblock - t.miniblocksToKeep
+	if miniblockToKeep <= 0 {
+		return nil // Nothing to trim
+	}
+
+	// Delete miniblocks in a single batch.
+	const batchSize = 1000
+	rows, err := tx.Query(
+		ctx,
+		t.store.sqlForStream(
+			`WITH rows_to_delete AS (
+				SELECT ctid
+				FROM {{miniblocks}}
+				WHERE stream_id = $1 AND seq_num < $2
+				LIMIT $3
+			)
+			DELETE FROM {{miniblocks}}
+			WHERE ctid IN (SELECT ctid FROM rows_to_delete)
+			RETURNING *`,
+			task.streamId,
+		),
+		task.streamId,
+		miniblockToKeep,
+		batchSize,
+	)
+	if err != nil {
+		return err
+	}
+
+	// Check if any rows were deleted
+	var deletedCount int
+	for rows.Next() {
+		deletedCount++
+	}
+	rows.Close()
+
+	// If we deleted a full batch, schedule another task
+	if deletedCount == batchSize {
+		t.scheduleTrimTask(task.streamId)
+	}
+
+	return nil
+}
+
+// scheduleTrimTask schedules a new trim task for a stream
+func (t *streamTrimmer) scheduleTrimTask(streamId StreamId) {
+	t.pendingTasksLock.Lock()
+	defer t.pendingTasksLock.Unlock()
+
+	// Check if there's already a pending task for this stream
+	if _, exists := t.pendingTasks[streamId]; exists {
+		return
+	}
+
+	// Mark the stream as having a pending task
+	t.pendingTasks[streamId] = struct{}{}
+
+	// Create a new task
+	task := trimTask{
+		streamId: streamId,
+	}
+
+	// Submit the task to the worker pool
+	t.workerPool.Submit(func() {
+		// Create a new context for the task
+		if err := t.processTrimTask(task); err != nil {
+			t.log.Errorw("Failed to process trim task",
+				"stream", task.streamId,
+				"err", err,
+			)
+		}
+
+		// Remove the pending task mark
+		t.pendingTasksLock.Lock()
+		delete(t.pendingTasks, streamId)
```

Comment:
In the following scenario:

task is picked up by the worker pool (`scheduleTrimTask` already returned), `processTrimTask` deletes batch size miniblocks and schedules a new trim task. `t.pendingTasks[streamId]` is not yet deleted (that happens afterwards) so the task is dropped due to a duplicate? And when `processTrimTask` returns the stream is deleted from pendingTasks on this line?